### ISSUE TITLE
ImageSegmentor - Export Shape file is not same as the Export Contour

### ISSUE THREAD

**Issue by hmbawa** [JzHuai0108/ImageSegmentor#1]
Hi Jianzhu,

I downloaded the ImageSegmentor and was able to compile and execute using Visual Studio 2010. However I have a problem. When I export the contour .tif file, I can see the smaller regions. However, when I export segmentation as the .shp I don't see smaller regions. I am using HierClust.

I also downloaded the executable file you put on csdn.net. It seems that results are different that the one produced by the build using source code.

Can you please help.

Thanks

Maz

**Comment by JzHuai0108** [comment]
I wish I could reply earlier. Thanks for your interest in using imageSegmentor.

I will try to find time to repeat and fix the issue. Because I may not do this soon, I would suggest you to upgrade the gdal libraries and see if this works. 

Right, I also put executables on csdn.net, but it is based on code a little bit older than that on github. If you find the results produced by the code on github is worse, please let me know.

**Comment by hmbawa** [comment]
Dear Huai,

Many thanks for the reply. I used on another computer the newer version of
GDAL and results are still different.
The executable you put on the csdn.net do produce the shape file however
the code that in on the github does not export to shape file and fails.
It seems that you were modifying or experimenting with the
CHC::Polygonize(CString
path) call as the definition of the projection dows not make sence

i.e. oSRS.SetGeogCS( "My geographic coordinate system",

My understanding is that when segmentation is done, the boundaries that I
see on the screen should be the Polygons in the .shp file once I export my
results as shape file.

Can you find out the older code for which you generated the executable on
csdn and email me, please !

Regards

Maz

mhbawa@gmail.com

**Comment by JzHuai0108** [comment]
You are right. Indeed the executable on csdn exports shapefiles into a subfolder called Polygon of the executable folder (really hacky!). 

Following your insight, I will try to fix the bug. Meanwhile, the source code to generate the executable on csdn is downloadable from http://download.csdn.net/download/pobudeyi/4472352. In case you have difficulty doing that, you may also find it at https://drive.google.com/open?id=0B_CdcisQWFcCU01LR1c5NEYtOUE.  Due to time limit, I did not compile and test it. If you encounter errors in doing that, please let me know.

**Comment by hmbawa** [comment]
Dear Huai,

I downloaded the code you sent and it seems that same as the one available
on github. I checked the part of the code doing the shape file export and
it is the same on both code bases.

Regards,

Maz

**Comment by hmbawa** [comment]
Dear Huai,

I used an image circle.tiff (with a simple ellipse and a rectangle)
attached here.
I used segmentation setting to be scale 10, shape 0.2, compactness 0.2 and
the weight of all three bands to be 1.
The results of the segmentation that I see on screen is as attached
result.jpg and the shape file exported is as attached Polygon.rar.

Please have a look.

Thanks,

Maz

**Comment by JzHuai0108** [comment]
You are absolutely right. The sections for exporting the shape file are identical in these code bases. And the executables built a long time ago from both code bases do export shape files. So I believe there may be some other issue causing the difficulty in exporting the shape file. 

Anyway, I need to overhaul the code for easier use.

Based on your latest reply, you said you attached the shape file result. So did you somehow make exporting the shape file work? 
And I am not seeing these attachments you mentioned, result.jpg, and Polygon.rar. Could you please send it again?

Thanks,

**Comment by hmbawa** [comment]
Many thanks for the reply Huai.

Please find attached the files.

I used the existing code to export by fixing some links and it it works
(though results are not right).

If you are still having difficulty in getting file, please send me some
other email address of yours.

Many thanks.

Maz

**Comment by JzHuai0108** [comment]
Oddly enough, I still don't see the attachments. The other email address I use is jianzhuhuai0108@gmail.com.
Thanks,
Jianzhu Huai

**Comment by JzHuai0108** [comment]
Good observation. I tried the program with the same parameter setting you used. Indeed it produced the same result. And the image only has two values 0 or 255 for R,G,B bands. By increasing the scale parameter (e.g., 80) or the spectral weight(i.e., color, e.g., 0.9), hierarchical clustering (HC) gets better results. Quad tree + HC does not work well either. In contrast, graph based segmentation works in this case, and hence, graph based segmentation +HC.

I believe, the problem is with the distance function used by hierarchical clustering which involves color and shape difference between regions. It encourages coherent regions in color by larger spectral weight and round regions in shape by larger compact weight.

Larger regions tend to have bigger difference in shape between each other, therefore they are not merged with a small scale parameter. 

Also, the region merging order also affects the result a little bit in this case. The Bayer dither matrix technique is used to ensure that all regions are growing evenly across the image.

**Comment by hmbawa** [comment]
A good explanation Huai.

It seems an interesting research area to further evolve the algorithms.

I believe you have some ideas to improve it too.

Will you be able to look the shape file export and results consistency
against what we see on screen. Also, try to run the segmentation with the
compiled version you put on csdn (the older one).

One thing I found and it is for your update, that as I am using Visual
Studio 10, the erase in the vector inside the iterator does not work with
the original code. We need to change:

S[ql].NPList.erase(ltemp);

with

lit = S[ql].NPList.erase(lit);

this way the std::vector erase will handle the change in the internal index
which gets changed in the case of erase.

Thanks.

Maz

**Comment by hmbawa** [comment]
Hi

I will try to get hold on the visual studio 6 and compile the original
 .dsw.

Thanks

Maz

**Comment by JzHuai0108** [comment]
You are right. The threshold may be made adaptive in the future. But now I am overloaded with dissertation. 

Oddly, the shape file does not include these outer boundaries and their associated edges.
Right. lit = S[ql].NPList.erase(lit); is the right syntax.

If I got time, I will try to write a cmake file for this project. Personally, I think Visual studio 6 is so old.

**Comment by hmbawa** [comment]
Many thanks for the reply again Huai,

I can understand that dissertation will take huge time and effort.

If you can find that very old code that exported the shape files properly,
I will wait once you finished the dissertation I will be happy working with
you to look into adaptive threshold.

Best Regards,

Maz
